File: mysql/boring/a_plain/plain.py
import csv
import time
import json
import psutil
import os
import MySQLdb

# ---------------- CONFIG ----------------
import sys

# Resolve paths relative to this script file so running from different CWDs works
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
CSV_FILE = os.path.normpath(os.path.join(BASE_DIR, '..', '..', '..', 'test-file.csv'))
RESULTS_DIR = os.path.normpath(os.path.join(BASE_DIR, '..', '..', '..', 'results'))
TABLE = "products_mysql_boring_plain"

BATCH_SIZE = 1000
TOTAL_ROWS = 1_000_000

# ---------------- METRICS ----------------
memory_usage = []
memory_spikes = []
cpu_usage = []
cpu_spikes = []

rows_inserted = 0
process = psutil.Process(os.getpid())


# ---------------- HELPERS ----------------
def record_metrics():
    mem = process.memory_info().rss / (1024 * 1024)  # MB
    cpu = process.cpu_percent(interval=None)

    memory_usage.append(mem)
    memory_spikes.append(max(mem, memory_spikes[-1] if memory_spikes else 0))

    cpu_usage.append(cpu)
    cpu_spikes.append(max(cpu, cpu_spikes[-1] if cpu_spikes else 0))


def insert_batch(cursor, batch):
    sql = f"""
        INSERT INTO {TABLE}
        (id, Name, Description, Brand, Category, Price, Currency, Stock,
         EAN, Color, Size, Availability, `InternalID`)
        VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
    """
    cursor.executemany(sql, batch)


# ---------------- MAIN ----------------
def main():
    global rows_inserted

    os.makedirs(RESULTS_DIR, exist_ok=True)

    if not os.path.exists(CSV_FILE):
        print(f"ERROR: CSV file not found at expected path: {CSV_FILE}")
        print("Make sure you run the script from the project root or that the test-file.csv exists.")
        sys.exit(1)

    db = MySQLdb.connect(
        host="localhost",
        user="root",
        passwd="",
        db="benchmark",
        autocommit=False
    )

    cursor = db.cursor()

    cursor.execute(f"TRUNCATE TABLE {TABLE}")
    db.commit()

    batch = []

    start_time = time.time()
    last_log = 0
    
    with open(CSV_FILE, newline='', encoding="utf-8") as csvfile:
        reader = csv.DictReader(csvfile)

        for row in reader:
            batch.append((
                row["Id"],
                row["Name"],
                row["Description"],
                row["Brand"],
                row["Category"],
                row["Price"],
                row["Currency"],
                row["Stock"],
                row["EAN"],
                row["Color"],
                row["Size"],
                row["Availability"],
                row["InternalID"],
            ))

            if len(batch) >= BATCH_SIZE:
                insert_batch(cursor, batch)
                db.commit()

                rows_inserted += len(batch)
                batch.clear()          # IMPORTANT: free references
                record_metrics()

                if rows_inserted - last_log >= 100_000:
                    print(f"Inserted {rows_inserted} rows")
                    last_log = rows_inserted

        if batch:
            insert_batch(cursor, batch)
            db.commit()
            rows_inserted += len(batch)
            batch.clear()
            record_metrics()

    total_time = time.time() - start_time
    rows_per_sec = rows_inserted / total_time

    result = {
        "db": "MySQL",
        "mode": "Boring",
        "variant": "Plain",
        "language": "Python",
        "total_rows": rows_inserted,
        "total_time_sec": round(total_time, 2),
        "rows_per_sec": round(rows_per_sec, 2),
        "peak_memory_mb": round(max(memory_usage), 2),
        "peak_cpu_percent": round(max(cpu_usage), 2),
        "memory_usage": memory_usage,
        "memory_spikes": memory_spikes,
        "cpu_usage": cpu_usage,
        "cpu_spikes": cpu_spikes,
    }

    with open(f"{RESULTS_DIR}/mysql_boring_a_plain_python.json", "w") as f:
        json.dump(result, f, indent=2)

    print("\n✅ DONE")
    print(f"Time: {total_time:.2f}s")
    print(f"Rows/sec: {rows_per_sec:.0f}")

    cursor.close()
    db.close()


if __name__ == "__main__":
    main()

File: mysql/boring/a_plain/plain.go
package main

import (
	"database/sql"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"time"

	_ "github.com/go-sql-driver/mysql"
	"github.com/shirou/gopsutil/v3/process"
)

const (
	csvFile    = "../../../test-file.csv"
	resultsDir = "../../../results"
	tableName  = "products_mysql_boring_plain"

	batchSize = 1000
	totalRows = 1000000
)

type Result struct {
	DB           string    `json:"db"`
	Mode         string    `json:"mode"`
	Variant      string    `json:"variant"`
	Language     string    `json:"language"`
	TotalRows    int       `json:"total_rows"`
	TotalTimeSec float64   `json:"total_time_sec"`
	RowsPerSec   float64   `json:"rows_per_sec"`
	PeakMemoryMB float64   `json:"peak_memory_mb"`
	PeakCPU      float64   `json:"peak_cpu_percent"`
	MemoryUsage  []float64 `json:"memory_usage"`
	MemorySpikes []float64 `json:"memory_spikes"`
	CPUUsage     []float64 `json:"cpu_usage"`
	CPUSpikes    []float64 `json:"cpu_spikes"`
}

func main() {
	_, thisFile, _, _ := runtime.Caller(0)

	baseDir := filepath.Dir(thisFile)

	resultsDir := filepath.Join(baseDir, resultsDir)

	os.MkdirAll(resultsDir, 0755)

	db, err := sql.Open("mysql", "root:@tcp(127.0.0.1:3306)/benchmark?parseTime=true")
	if err != nil {
		panic(err)
	}
	defer db.Close()

	db.Exec("TRUNCATE TABLE " + tableName)

	csvFile := filepath.Join(baseDir, csvFile)

	file, err := os.Open(csvFile)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	reader.FieldsPerRecord = -1

	// skip header
	reader.Read()

	proc, _ := process.NewProcess(int32(os.Getpid()))

	start := time.Now()
	rowsInserted := 0
	lastLog := 0

	var memoryUsage, memorySpikes, cpuUsage, cpuSpikes []float64

	batch := make([][]string, 0, batchSize)

	recordMetrics := func() {
		memInfo, _ := proc.MemoryInfo()
		cpu, _ := proc.CPUPercent()

		mem := float64(memInfo.RSS) / 1024 / 1024

		memoryUsage = append(memoryUsage, mem)
		cpuUsage = append(cpuUsage, cpu)

		if len(memorySpikes) == 0 || mem > memorySpikes[len(memorySpikes)-1] {
			memorySpikes = append(memorySpikes, mem)
		} else {
			memorySpikes = append(memorySpikes, memorySpikes[len(memorySpikes)-1])
		}

		if len(cpuSpikes) == 0 || cpu > cpuSpikes[len(cpuSpikes)-1] {
			cpuSpikes = append(cpuSpikes, cpu)
		} else {
			cpuSpikes = append(cpuSpikes, cpuSpikes[len(cpuSpikes)-1])
		}
	}

	insertBatch := func(batch [][]string) {
		tx, _ := db.Begin()
		stmt, _ := tx.Prepare(`
			INSERT INTO ` + tableName + `
			(Id, Name, Description, Brand, Category, Price, Currency, Stock,
			 EAN, Color, Size, Availability, InternalID)
			VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
		`)
		defer stmt.Close()

		for _, r := range batch {
			price, _ := strconv.ParseFloat(r[5], 64)
			stock, _ := strconv.Atoi(r[7])

			stmt.Exec(
				r[0], r[1], r[2], r[3], r[4],
				price, r[6], stock,
				r[8], r[9], r[10], r[11], r[12],
			)
		}

		tx.Commit()
	}

	for {
		row, err := reader.Read()
		if err != nil {
			break
		}

		batch = append(batch, row)

		if len(batch) >= batchSize {
			insertBatch(batch)
			rowsInserted += len(batch)
			batch = batch[:0]

			recordMetrics()

			if rowsInserted-lastLog >= 100000 {
				fmt.Println("Inserted", rowsInserted)
				lastLog = rowsInserted
			}
		}
	}

	if len(batch) > 0 {
		insertBatch(batch)
		rowsInserted += len(batch)
		recordMetrics()
	}

	elapsed := time.Since(start).Seconds()
	rowsPerSec := float64(rowsInserted) / elapsed

	var peakMem, peakCPU float64
	for _, m := range memoryUsage {
		if m > peakMem {
			peakMem = m
		}
	}
	for _, c := range cpuUsage {
		if c > peakCPU {
			peakCPU = c
		}
	}

	result := Result{
		DB:           "MySQL",
		Mode:         "Boring",
		Variant:      "Plain",
		Language:     "Go",
		TotalRows:    rowsInserted,
		TotalTimeSec: elapsed,
		RowsPerSec:   rowsPerSec,
		PeakMemoryMB: peakMem,
		PeakCPU:      peakCPU,
		MemoryUsage:  memoryUsage,
		MemorySpikes: memorySpikes,
		CPUUsage:     cpuUsage,
		CPUSpikes:    cpuSpikes,
	}

	f, _ := os.Create(resultsDir + "/mysql_boring_a_plain_go.json")
	json.NewEncoder(f).Encode(result)
	f.Close()

	fmt.Println("\n✅ DONE")
	fmt.Printf("Time: %.2fs\n", elapsed)
	fmt.Printf("Rows/sec: %.0f\n", rowsPerSec)
	fmt.Printf("Peak RAM: %.2f MB\n", peakMem)
	fmt.Printf("Peak CPU: %.2f%%\n", peakCPU)

	runtime.GC()
}
